### ISSUE TITLE
Force grid_division to integer values

### ISSUE THREAD

**Issue by AbcAeffchen** [goat1000/SVGGraph#27]
Hi,

I try to port some plots from jpgraph to SVGGraph. It looks pretty good at the moment but I cannot get the grid division right. I have some line plots where the y-values are all integers. Sadly SVGGraph decides that 1.9 is a good division...I found the option `decimal_digits` but this can only hide decimal digits and leads to strange behavior due to rounding.

I think the problem here is, that the code that calculates the division hat not the option to change the maximum of the grid. It would be nice to have an option to force an integer division. One could just round the division up and check the number if divisions needed and maybe increase the max-value of the plot a little bit.

Best,
Alex

**Comment by goat1000** [comment]
Hi Alex,

There is already an option for this, but it might not have the most obvious name: `minimum_units_y`.

The docs for it are here: http://www.goat1000.com/svggraph-docs/minimum_units_y

Thanks,
Graham

**Comment by AbcAeffchen** [comment]
Hi Graham,

Ah very nice. Thanks a lot :)
I think I wouldn't have found this. 

Best,
Alex

**Comment by AbcAeffchen** [comment]
Hi Graham,

Sorry to reopen this issue. At first `minimum_units_y` seemed to fix my problem and the link you provided says:
>Setting the option to 1 will make SVGGraph use whole numbers for the axis labels.

I tried it and it worked fine. But then I had a test dataset that forced SVGGraph to use a very big division again with non-integers.

So as the name of this setting suggests it is a minimum unit size. But floating point numbers above this minimum can still be used.

Best,
Alex

**Comment by goat1000** [comment]
Hi Alex

I'll have a look at this tomorrow - numbers like 1.9 don't really sound like a good choice for divisions, so I'll try to fix the algorithm.

Thanks,
Graham

**Comment by AbcAeffchen** [comment]
Thanks Graham. No hurry. Let me know if I should look into something.

**Comment by goat1000** [comment]
I've made an update in ce1b66f that should improve things - it can still pick values like 0.1 and 2.5 as divisions, but should not pick 0.9 or 1.9 any more.

Anyone who wants to use divisions like that can still specify them using the `grid_division_*` options.

**Comment by AbcAeffchen** [comment]
Thanks a lot. I will look into this as soon as possible.

**Comment by AbcAeffchen** [comment]
I tested your changes and now I get 2 instead of 1.9 right away. So this is fine.
![12_csp_0](https://user-images.githubusercontent.com/5396140/62327140-1ae57580-b4b0-11e9-8120-644bff3affec.png)

But if I increase `minimum_grid_spacing` from 15 (default) to 20, 2 doesn't work anymore and the only other integer divisor of 38 is 19, so it chooses 9.5. like here

![12_csp_1](https://user-images.githubusercontent.com/5396140/62327304-8d565580-b4b0-11e9-8f8e-4eb9af5f503e.png)

Here it would be nice if the algorithm just chose 3 and set a maximum to 39, or 4 an set it to 40 and scales everything accordingly. 

If you like I try to find some time to implement this but it could take a while since I also need to read into the code first.

**Comment by goat1000** [comment]
I'll take another look at this tomorrow - there are a couple of different places where I can see room for improvement, I just have to make sure I don't break really large or really small or negative numbers.

Thanks,
Graham

**Comment by goat1000** [comment]
I've just uploaded another update, 4f280ad. Last time I only changed the function that decided which values were acceptable as divisions, but this time I've changed the division selection algorithm itself.

The previous version would choose the first acceptable division it found, this time it creates a list of all the acceptable divisions and chooses the one with the best "score". It also allows the scale to grow by up to 1/5, which makes it work a lot better with prime numbers.

The previous algorithm was not very good with negative values either and this one seems a lot better.
